### PR TITLE
Add cachedIOWith

### PR DESCRIPTION
**Pull request by Arguggi**
I needed to cache an IO action and only update it when the day changes, hence the changes.

The old function still has the same behavior so a simple bump to 1.0.1.0 should be enough if I'm not mistaken.

Let me know if there is something I should fix before you merge the changes.

### CODE REVIEW

**Comment by Arguggi** [comment]
I fixed the cabal settings for the test executable since the cached-io library wasn't in the build-depends.

I'm not 100% sure of the MonadIO change but since I'm using it in my pet project I figured I would add it to the pull request, feel free to drop this last commit if you prefer to continue using IO instead of depending on transformers obviously.

**Comment by taphu** [comment]
@Arguggi sorry for the delay. I've been on vacation. I will try to take a look at this as soon as I can, which should be today or tomorrow.

File: src/Control/Concurrent/CachedIO.hs [hunk 1]
```diff
@@ -14,8 +15,29 @@ data State = Uninitialized | Initializing
...
-cachedIO interval io = do
+cachedIO :: NominalDiffTime -- ^ Number of seconds before refreshing cache
+         -> IO a            -- ^ IO action to cache
+         -> IO (IO a)
+cachedIO interval = cachedIOWith (intervalPassed interval)
+
+-- | Check if `interval` seconds have passed from the starting time
+intervalPassed :: NominalDiffTime  -- ^ Seconds
+               -> UTCTime          -- ^ Last time cache was updated
+               -> UTCTime          -- ^ Now
```
Comment (MHova) [inline src/Control/Concurrent/CachedIO.hs]: Calling this "now" would be selling this function short as it calculates whether the interval has passed between any two arbitrary times :smile:

File: src/Control/Concurrent/CachedIO.hs [hunk 2]
```diff
@@ -6,6 +6,7 @@ module Control.Concurrent.CachedIO (
 import Control.Concurrent.STM (atomically, newTVar, readTVar, writeTVar, retry)
 import Control.Monad (join)
 import Data.Time.Clock (NominalDiffTime, addUTCTime, getCurrentTime, UTCTime)
+import Control.Monad.IO.Class
```
Comment (MHova) [inline src/Control/Concurrent/CachedIO.hs]: Explicit import?

File: src/Control/Concurrent/CachedIO.hs [hunk 3]
```diff
@@ -33,40 +36,49 @@ intervalPassed interval lastUpdated now = addUTCTime interval lastUpdated > now
...
                                              --   the cache is considered still fresh
                                              --   and returns the cached IO action
-             -> m a                          -- ^ action to cache
-             -> m (m a)
+             -> t a                          -- ^ action to cache
+             -> m (t a)
 cachedIOWith test io = do
   initTime <- liftIO getCurrentTime
-  cachedT <- liftIO (atomically (newTVar (initTime, Left Uninitialized)))
+  cachedT <- liftIO . atomically $ newTVar (initTime, Left Uninitialized)
```
Comment (MHova) [inline src/Control/Concurrent/CachedIO.hs]: I'll defer to @taphu on style issues

File: src/Control/Concurrent/CachedIO.hs [hunk 4]
```diff
@@ -14,8 +15,29 @@ data State = Uninitialized | Initializing
...
+--
+-- The outer IO is responsible for setting up the cache. Use the inner one to
+-- either get the cached value or refresh
+cachedIOWith :: (UTCTime -> UTCTime -> Bool) -- ^ Test function:
+                                             --   If 'test' 'lastUpdated' 'now' returns 'True'
+                                             --   the cache is considered still fresh
+                                             --   and returns the cached IO action
+             -> IO a                         -- ^ IO action to cache
+             -> IO (IO a)
+cachedIOWith test io = do
```
Comment (MHova) [inline src/Control/Concurrent/CachedIO.hs]: I had to reread the later code a couple of times (despite being the original author) to figure out wtf was going on. Perhaps we should rename `test` to something like `isCacheStillFresh` to help out the next poor soul.

File: src/Control/Concurrent/CachedIO.hs [hunk 5]
```diff
@@ -33,40 +36,49 @@ intervalPassed interval lastUpdated now = addUTCTime interval lastUpdated > now
...
-          writeTVar cachedT (now, Right value)
-          return $ refreshCache now cachedT
-        -- The cache is uninitialized. Mark the cache as initializing to block other
+        (lastUpdated, Right value)
+          | test lastUpdated now -> return (return value)
+          -- There's data in the cache, but it's stale. Update the cache status
+          -- to prevent a second thread from also executing the IO action. The second
+          -- thread would get the stale data instead. Refresh the cache and return.
+          | otherwise -> do
+            writeTVar cachedT (now, Left Updating)
```
Comment (MHova) [inline src/Control/Concurrent/CachedIO.hs]: Changing this to `Left Updating` removes the "another thread will just get stale data" behavior. Is this change necessary for this PR?

See: https://github.com/SumAll/haskell-cached-io/commit/4050d8212875c5a548dab472655a719539b58607

File: src/Control/Concurrent/CachedIO.hs [hunk 6]
```diff
@@ -33,40 +36,49 @@ intervalPassed interval lastUpdated now = addUTCTime interval lastUpdated > now
...
       liftIO (atomically (writeTVar cachedT (now, Right newValue)))
       liftIO (return newValue)
+
+-- We only want to restore the state of the TVar so that any other thread that was waiting
+-- will wake up and will see the Uninitialized state forcing an update (The cache was stale/empty
+-- since someone ran the action, so we don't put back the value the cache had before).
+restoreUninitialized :: (MonadIO m)
+                     => UTCTime
+                     -> TVar (UTCTime, Either State a) -> m ()
+restoreUninitialized now cachedT = liftIO (atomically (writeTVar cachedT (now, Left Uninitialized)))
```
Comment (MHova) [inline src/Control/Concurrent/CachedIO.hs]: Should we instead retain the previous `lastUpdated` and `Either value`?

File: src/Control/Concurrent/CachedIO.hs (hunk 3 above)
Comment (taphu) [inline src/Control/Concurrent/CachedIO.hs]: I'm cool with the style change, though I typically prefer that no-op/style commits be separate from "real" commits.

Comment (Arguggi) [inline src/Control/Concurrent/CachedIO.hs]: Sorry I'm more used to a more point free style and this was not intended

File: src/Control/Concurrent/CachedIO.hs (hunk 5 above)
Comment (Arguggi) [inline src/Control/Concurrent/CachedIO.hs]: I'm not sure I follow. Now when the cache is stale or uninitialized the first thread that requests the value will block all others by setting (Left Updating) (so others will retry) and once that thread finishes it either resets the cache (it actually threw an exception) or the new value will be in the cache.

 We want this behavior right? We don't want other threads to get stale data.

(Yes I should probably have opened another PR, sorry)

File: src/Control/Concurrent/CachedIO.hs (hunk 2 above)
Comment (Arguggi) [inline src/Control/Concurrent/CachedIO.hs]: I'll add MonadIO

File: src/Control/Concurrent/CachedIO.hs (hunk 4 above)
Comment (Arguggi) [inline src/Control/Concurrent/CachedIO.hs]: I agree the test name is bad, I'll rename it. The comments should help a little bit though.

File: src/Control/Concurrent/CachedIO.hs (hunk 6 above)
Comment (Arguggi) [inline src/Control/Concurrent/CachedIO.hs]: My reasoning was that since a thread was running the action the data was stale so all other threads should just get the new value. Putting Left Uninitialized or whatever was the Either value before should be the same since any thread that comes after the one that threw an exception should decide to update the cache anyway (unless the test condition is something strange like update only if the minutes are odd)

File: src/Control/Concurrent/CachedIO.hs (hunk 5 above)
Comment (MHova) [inline src/Control/Concurrent/CachedIO.hs]: I think this library has had the "stale data" behavior since its inception. Certainly both "stale data" and "block/retry" have their own merits and are both "correct" in their own ways. I'm hesitant to change the existing contract in this subtle way, though.

Can you exclude this particular change from this PR and maybe open another PR about this? Perhaps we should allow the client to specify whether they want blocking or non-blocking behavior for refreshes (and default to non-blocking).

File: src/Control/Concurrent/CachedIO.hs (hunk 6 above)
Comment (MHova) [inline src/Control/Concurrent/CachedIO.hs]: I think the safer, more defensive way is to retain the existing data. This would trigger an update anyway given a standard `isCacheStillFresh` condition, but would also do the right thing given a bizarre condition.

File: src/Control/Concurrent/CachedIO.hs (hunk 3 above)
Comment (MHova) [inline src/Control/Concurrent/CachedIO.hs]: For the record, I prefer the point free style as well =)

File: src/Control/Concurrent/CachedIO.hs (hunk 5 above)
Comment (Arguggi) [inline src/Control/Concurrent/CachedIO.hs]: I think the old behaviour ("stale data") is broken in a subtle way, I'll remove the changes from this PR and open a new one and try to explain there.

**Comment by Arguggi** [comment]
I rebased and removed the commits with the "stale cache" / "block/retry" changes
